### PR TITLE
Switch to prisma

### PR DESCRIPTION
**Pull request by Huskydog9988**
An attempt to switch map sync over from typeorm to prisma.

Rn a migration is needed to on the users side to switch from the old typeorm system to prisma. We can solve this by renaming certain things into the snake case previously used. This is a problem as we need a way to automate running migrations in the server before the server starts without much impact. There have already been some [preposed solutions](https://dev.to/awohletz/running-prisma-migrate-in-an-electron-app-1ehm), but are slightly hacky as they require hardcoding some stuff, plus running cmds directly with node.

### CODE REVIEW

File: server/prisma/schema.prisma [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+datasource db {
+  provider = "sqlite"
+  url      = env("DATABASE_URL")
+}
+
+model ChunkData {
+  hash        Bytes         @id
+  version     Int
+  data        Bytes
```
Comment (Gjum) [inline server/prisma/schema.prisma:13]: missing world/chunkX/chunkZ/timestamp, will be in next commit I presume but adding this comment so I remember

File: server/src/MapChunk.ts [hunk 2]
```diff
@@ -67,12 +68,54 @@ export class PlayerChunkDB extends BaseEntity implements PlayerChunk {
...
 	static async store(map_chunk: PlayerChunk) {
 		// TODO if PlayerChunk exists, and holds last reference to old hash, delete ChunkData at old hash
-		await ChunkDataDB.upsert(map_chunk.data, {
-			conflictPaths: ChunkDataDB.primaryCols,
-			skipUpdateIfNoValuesChanged: true,
+
+		const chunkData = {
+			data: map_chunk.data.data,
+			hash: map_chunk.data.hash,
+			version: map_chunk.data.version,
```
Comment (Gjum) [inline server/src/MapChunk.ts:81]: needs world/chunk_x/chunk_z/timestamp for time travel

File: README.md [hunk 3]
```diff
@@ -30,16 +30,14 @@ You can control who has access to a Sync Server by editing its `allowed-users.tx
...
-    - `docker compose up map-sync -d`
-    - To stop the headless server: `docker compose down map-sync`
+4. Clone our code.
+   - `git clone https://github.com/CivPlatform/map-sync.git`
+5. Change your working directory.
+   - `cd map-sync/`
+6. To run the server with interactive prompt:
+   - `docker compose run --rm -it -p 12312:12312 map-sync`
+   - To stop the interactive prompt: hit ctrl-c twice
+7. To run the server headless: - `docker compose up map-sync -d` - To stop the headless server: `docker compose down map-sync`
```
[the remark this answers is not in the document]
Comment (Huskydog9988) [reply README.md:40]: yea i have prettier installed in vs, i think it may have messed with formating

File: server/prisma/schema.prisma [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+model ChunkData {
+  hash      Bytes  @id
+  version   Int
+  data      Bytes
+  world     String
+  chunkX    Int
+  chunkZ    Int
+  timestamp BigInt
```
Comment (Gjum) [inline server/prisma/schema.prisma:17]: I believe we should store the timestamp as an integer, in seconds. BigInts/bigints are unreliable in NodeJS in my experience (see Discord discussion). What do you think?

File: server/src/MapChunk.ts [hunk 5]
```diff
@@ -67,13 +68,60 @@ export class PlayerChunkDB extends BaseEntity implements PlayerChunk {
...
-			skipUpdateIfNoValuesChanged: true,
+
+		// create object to store all data needed for chunk data
+		const chunkData = {
+			data: map_chunk.data.data,
+			hash: map_chunk.data.hash,
+			version: map_chunk.data.version,
+			chunkX: map_chunk.chunk_x,
+			chunkZ: map_chunk.chunk_z,
+			timestamp: map_chunk.ts,
```
Comment (Gjum) [inline server/src/MapChunk.ts:85]: If we store the timestamp in seconds in the DB, it needs to be converted (multiply/divide by 1000) here and in other places.

File: server/src/MapChunk.ts [hunk 6]
```diff
@@ -67,13 +68,60 @@ export class PlayerChunkDB extends BaseEntity implements PlayerChunk {
...
+					},
+				},
+			},
+			update: {
+				...playerData,
+				timestamp: map_chunk.ts,
+				chunkData: {
+					upsert: {
+						create: chunkData,
+						update: chunkData,
```
Comment (Gjum) [inline server/src/MapChunk.ts:120]: We should only set the timestamp if this ChunkData doesn't already exist, as that's the most accurate time when the chunk last changed. It should be the time this ChunkData was first created; actually that would also be a good column name.

**Review by Gjum — CHANGES_REQUESTED** [review]

File: server/src/MapChunk.ts (hunk 6 above)
Comment (Huskydog9988) [reply server/src/MapChunk.ts:120]: should be fixed in 92f0417

**Comment by Huskydog9988** [comment]
Gonna post this inspiration for myself, umami seems to have figured out prisma [migrations on the fly](https://github.com/umami-software/umami/blob/master/scripts/check-db.js).

**Pull request closed without merging**